src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +28,99 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(base_segment);
+    const auto values = value_segment->values();
+
+    std::set<T> temp_value_set;
+
+    for (auto& value : values) {
+      temp_value_set.insert(value);
+    }
+
+    const auto set_size = temp_value_set.size();
+    if (set_size <= std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector =
+          std::dynamic_pointer_cast<BaseAttributeVector>(std::make_shared<FixedSizeAttributeVector<uint8_t>>());
+    } else if (set_size <= std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector =
+          std::dynamic_pointer_cast<BaseAttributeVector>(std::make_shared<FixedSizeAttributeVector<uint16_t>>());
+    } else {
+      _attribute_vector =
+          std::dynamic_pointer_cast<BaseAttributeVector>(std::make_shared<FixedSizeAttributeVector<uint32_t>>());
+    }
+
+    _dictionary = std::make_shared<std::vector<T>>(temp_value_set.cbegin(), temp_value_set.cend());
+
+    for (auto position = 0UL; position < values.size(); position++) {
+      const auto element_it = std::lower_bound(_dictionary->cbegin(), _dictionary->cend(), values[position]);
+      _attribute_vector->set(position, static_cast<ValueID>(element_it - _dictionary->cbegin()));
+    }
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    return _dictionary->operator[](_attribute_vector->get(chunk_offset));
+  }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const { return _dictionary->operator[](_attribute_vector->get(chunk_offset)); }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant&) override;
+  void append(const AllTypeVariant&) override { throw std::runtime_error("Dictionary segments are immutable."); }
```

Comment:
Maybe use an Assert?
`Assert(true, "Dictionary segments are immutable.");`

Reply:
Assert(True, ...) just to print something... We are not in Timbuktu here.